tests/drift_kinetic_push/drift_kinetic_grid_ex2.cpp

```diff
@@ -0,0 +1,342 @@
+#include "drift_kinetic_push.h"
+
+static constexpr char help[] =
+  "Test drift-kinetic Esirkepov vs analytical: comparison of fields and trajectories\n"
+  "calculated by drift_kinetic_push on analytical fields vs on grid filled with\n"
+  "analytical fields. Controls boundary violations and particle displacement < cell size.\n";
+
+constexpr Vector3R E0(0.0, 1.0, -1.0);
+constexpr Vector3R B0(0.0, 0.0, 1.0);
+constexpr Vector3R gradB0(0.0, 0.0, 0.0);
+
+struct TestStatistics {
+  PetscReal max_field_error_E = 0.0;
+  PetscReal max_field_error_B = 0.0;
+  PetscReal max_field_error_gradB = 0.0;
+  PetscReal max_position_error = 0.0;
+  PetscReal max_displacement_per_step = 0.0;
+  PetscInt total_iterations_analytical = 0;
+  PetscInt total_iterations_grid = 0;
+  PetscInt boundary_violations = 0;
+  PetscInt cell_size_violations = 0;
+  Vector3R final_position_analytical;
+  Vector3R final_position_grid;
+  PetscReal simulation_time = 0.0;
+  PetscInt total_steps = 0;
+};
+
+void get_analytical_fields(const Vector3R&, const Vector3R&, Vector3R& E_p, Vector3R& B_p, Vector3R& gradB_p)
+{
+  E_p = E0;
+  B_p = B0;
+  gradB_p = gradB0;
+}
+
+PetscErrorCode initialize_grid_fields(DM da, Vec E_vec, Vec B_vec, Vec gradB_vec)
+{
+  PetscFunctionBeginUser;
+
+  Vector3R*** E_arr, ***B_arr, ***gradB_arr;
+
+  PetscCall(DMDAVecGetArrayWrite(da, E_vec, &E_arr));
+  PetscCall(DMDAVecGetArrayWrite(da, B_vec, &B_arr));
+  PetscCall(DMDAVecGetArrayWrite(da, gradB_vec, &gradB_arr));
+
+  Vector3I start, size;
+  PetscCall(DMDAGetCorners(da, REP3_A(&start), REP3_A(&size)));
+
+  for (PetscInt k = start[Z]; k < start[Z] + size[Z]; k++) {
+    for (PetscInt j = start[Y]; j < start[Y] + size[Y]; j++) {
+      for (PetscInt i = start[X]; i < start[X] + size[X]; i++) {
+        E_arr[k][j][i] = E0;
+        B_arr[k][j][i] = B0;
+        gradB_arr[k][j][i] = gradB0;
+      }
+    }
+  }
+
+  PetscCall(DMDAVecRestoreArrayWrite(da, E_vec, &E_arr));
+  PetscCall(DMDAVecRestoreArrayWrite(da, B_vec, &B_arr));
+  PetscCall(DMDAVecRestoreArrayWrite(da, gradB_vec, &gradB_arr));
+
```

Comment:
If fields are constant ($\nabla B_p = 0$), you can simply set each component

```suggestion
  PetscCall(VecStrideSet(E_vec, X, E0[X]));
  PetscCall(VecStrideSet(E_vec, Y, E0[Y]));
  PetscCall(VecStrideSet(E_vec, Z, E0[Z]));
  
  PetscCall(VecStrideSet(B_vec, X, B0[X]));
  PetscCall(VecStrideSet(B_vec, Y, B0[Y]));
  PetscCall(VecStrideSet(B_vec, Z, B0[Z]));
```

see https://petsc.org/release/manualpages/Vec/VecStrideSet/